packages/apollo-server-core/src/plugin/schemaReporting/index.ts

```diff
@@ -0,0 +1,149 @@
+import os from 'os';
+import { ApolloServerPlugin } from 'apollo-server-plugin-base';
+import { v4 as uuidv4 } from 'uuid';
+import { printSchema } from 'graphql';
+import { SchemaReporter } from './schemaReporter';
+import createSHA from '../../utils/createSHA';
+
+export interface ApolloServerPluginSchemaReportingOptions {
+  /**
+   * The schema reporter waits before starting reporting.
+   * By default, the report waits some random amount of time between 0 and 10 seconds.
+   * A longer interval leads to more staggered starts which means it is less likely
+   * multiple servers will get asked to upload the same schema.
+   *
+   * If this server runs in lambda or in other constrained environments it would be useful
+   * to decrease the schema reporting max wait time to be less than default.
+   *
+   * This number will be the max for the range in ms that the schema reporter will
+   * wait before starting to report.
+   */
+  initialDelayMaxMs?: number;
+  /**
+   * Override the reported schema that is reported to the Apollo registry. This
+   * schema does not go through any normalizations and the string is directly
+   * sent to the Apollo registry. This can be useful for comments or other
+   * ordering and whitespace changes that get stripped when generating a
+   * `GraphQLSchema`.
+   *
+   * **If you pass this option to this plugin, you should explicitly configure
+   * `ApolloServerPluginUsageReporting` and pass the same value to its
+   * `overrideReportedSchema` option.** This ensures that the schema ID
+   * associated with requests reported by the usage reporting plugin matches the
+   * schema ID that this plugin reports. For example:
+   *
+   * ```js
+   * new ApolloServer({
+   *   plugins: [
+   *     ApolloServerPluginSchemaReporting({overrideReportedSchema: schema}),
+   *     ApolloServerPluginUsageReporting({overrideReportedSchema: schema}),
+   *   ],
+   * })
+   * ```
+   */
+  overrideReportedSchema?: string;
+  /**
+   * The URL to use for reporting schemas. Primarily for testing and internal
+   * Apollo use.
+   */
+  endpointUrl?: string;
+}
+
+export function ApolloServerPluginSchemaReporting(
+  {
+    initialDelayMaxMs,
+    overrideReportedSchema,
+    endpointUrl,
+  }: ApolloServerPluginSchemaReportingOptions = Object.create(null),
+): ApolloServerPlugin {
+  const bootId = uuidv4();
+
+  return {
+    __internal_plugin_id__() {
+      return 'SchemaReporting';
+    },
+    async serverWillStart({ apollo, schema, logger }) {
+      const { key, graphId } = apollo;
+      if (!key) {
+        throw Error(
+          'To use ApolloServerPluginSchemaReporting, you must provide an Apollo API ' +
+            'key, via the $APOLLO_KEY environment variable or via `new ApolloServer({apollo: {key})`',
```

Comment:
That's reasonable.

I think that when I think about it, which I wasn't necessarily doing, I like using `$` in contexts where we're being brief and trying to avoid the phrase `environment variable`, but it's redundant together with it. I've gone back and applied that rule. If you want me to just go all the way and replace all $FOO with "the FOO environment variable", happy to do so before release.